File: backend/app/schemas/market.py
"""Market data schemas for API."""

from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel, ConfigDict


class QuoteResponse(BaseModel):
    """Individual stock quote."""

    model_config = ConfigDict(from_attributes=True)

    symbol: str
    price: Decimal
    change: Decimal | None
    change_pct: Decimal | None
    volume: int | None
    name: str | None
    sector: str | None
    updated_at: datetime


class SectorData(BaseModel):
    """Sector performance for heatmap."""

    sector: str
    change_pct: Decimal
    market_cap: int | None


class SectorHeatmapResponse(BaseModel):
    """Sector heatmap response."""

    sectors: list[SectorData]
    updated_at: datetime


class TopMover(BaseModel):
    """Top mover stock."""

    symbol: str
    name: str | None
    price: Decimal
    change_pct: Decimal
    volume: int | None


class TopMoversResponse(BaseModel):
    """Top movers response."""

    gainers: list[TopMover]
    losers: list[TopMover]
    most_active: list[TopMover]
    updated_at: datetime


class IndexData(BaseModel):
    """Market index data."""

    symbol: str
    name: str
    value: Decimal
    change: Decimal | None
    change_pct: Decimal | None


class MarketIndicesResponse(BaseModel):
    """Market indices response."""

    indices: list[IndexData]
    updated_at: datetime


class SymbolSearchResult(BaseModel):
    """Symbol search result."""

    symbol: str
    name: str
    sector: str | None


class SymbolSearchResponse(BaseModel):
    """Symbol search response."""

    results: list[SymbolSearchResult]

File: frontend-react/src/api/market.ts
/**
 * Market data API calls
 */
import apiClient from './client';

export interface Quote {
  symbol: string;
  price: string;
  change: string | null;
  change_pct: string | null;
  volume: number | null;
  name: string | null;
  sector: string | null;
  updated_at: string;
}

export interface SectorData {
  sector: string;
  change_pct: string;
  market_cap: number | null;
}

export interface SectorHeatmapResponse {
  sectors: SectorData[];
  updated_at: string;
}

export interface TopMover {
  symbol: string;
  name: string | null;
  price: string;
  change_pct: string;
  volume: number | null;
}

export interface TopMoversResponse {
  gainers: TopMover[];
  losers: TopMover[];
  most_active: TopMover[];
  updated_at: string;
}

export interface IndexData {
  symbol: string;
  name: string;
  value: string;
  change: string | null;
  change_pct: string | null;
}

export interface MarketIndicesResponse {
  indices: IndexData[];
  updated_at: string;
}

export interface SymbolSearchResult {
  symbol: string;
  name: string;
  sector: string | null;
}

export interface SymbolSearchResponse {
  results: SymbolSearchResult[];
}

/**
 * Get quote for a specific symbol
 */
export const getQuote = async (symbol: string): Promise<Quote> => {
  const response = await apiClient.get<Quote>(`/market/quote/${symbol}`);
  return response.data;
};

/**
 * Get sector heatmap data
 */
export const getSectorHeatmap = async (): Promise<SectorHeatmapResponse> => {
  const response = await apiClient.get<SectorHeatmapResponse>('/market/sectors');
  return response.data;
};

/**
 * Get top movers
 */
export const getTopMovers = async (limit = 10): Promise<TopMoversResponse> => {
  const response = await apiClient.get<TopMoversResponse>('/market/movers', {
    params: { limit },
  });
  return response.data;
};

/**
 * Get market indices
 */
export const getMarketIndices = async (): Promise<MarketIndicesResponse> => {
  const response = await apiClient.get<MarketIndicesResponse>('/market/indices');
  return response.data;
};

/**
 * Search for symbols
 */
export const searchSymbols = async (
  query: string,
  limit = 20
): Promise<SymbolSearchResponse> => {
  const response = await apiClient.get<SymbolSearchResponse>('/market/search', {
    params: { q: query, limit },
  });
  return response.data;
};
